src/main/java/net/minecraftforge/logging/CrashReportAnalyser.java

```diff
@@ -0,0 +1,280 @@
+        }
+        catch(Throwable t) {
+            LOGGER.error("Failed to analyse crash report!", t);
+        }
+    }
+
+    /**
+     * Checks the stacktrace of the given throwable and all its children for any occurrence of the base package names cached in the basePackageModCache,
+     * including mixin classes.
+     * **/
+    private static void scanThrowable(Throwable throwable) {
+        scanStacktrace(throwable.getStackTrace());
+
+        if (throwable.getCause() != null && throwable.getCause() != throwable) {
+            scanThrowable(throwable.getCause());
+        }
+    }
+
+    /**
+     * Checks the given stacktrace for any occurrence of the base package names cached in the basePackageModCache,
+     * including mixin classes.
+     * **/
+    private static void scanStacktrace(StackTraceElement[] stackTrace) {
+        for (StackTraceElement stackTraceElement : stackTrace) {
+            identifyByClass(stackTraceElement);
+            identifyByMixins(stackTraceElement);
+        }
+    }
+
+
+    /**
+     * Iterates over all loaded mods, resolving and caching their base package names with the corresponding {@link IModInfo}.
+     * <ul>
+     *     <li>For forge this would be: "net.minecraftforge"</li>
+     *     <li>For minecraft this would be:</li>
+     *     <ul>
+     *         <li>"net.minecraft"</li>
+     *         <li>"com.mojang"</li>
+     *         <li>"mcp.client"</li>
+     *     </ul>
+     * </ul>
+     * **/
+    public static void cacheModList() {
+        ModList.get().getMods().forEach(iModInfo -> {
+            //Don't cache minecraft or forge as they will always be included in the stacktrace
+            if (!iModInfo.getModId().equals("forge") && !iModInfo.getModId().equals("minecraft")) {
+                ModFileScanData scanData = iModInfo.getOwningFile().getFile().getScanResult();
+                Collection<String> commonClassPaths = resolveCommonClassPaths(scanData.getClasses());
+                commonClassPaths.forEach(s -> BASE_PACKAGE_MOD_CACHE.put(s, iModInfo));
+            }
+        });
+    }
+
+    private static void identifyByClass(StackTraceElement stackTraceElement) {
+        blameIfPresent(stackTraceElement);
+    }
+
+    private static void identifyByMixins(StackTraceElement stackTraceElement) {
+        IMixinInfo mixinInfo = getMixinInfo(stackTraceElement);
+
+        if (mixinInfo != null) {
+            String elementAsString = stackTraceElement.toString();
+            String mixinClassName = mixinInfo.getClassName();
+            List<String> targetClasses = mixinInfo.getTargetClasses();
+            blameIfPresent(
+                    "Mixin class: " + mixinClassName +
+                            "\n\t\tTarget" + (targetClasses.size() == 1 ? ": " + targetClasses.get(0).replaceAll("/", ".") : "s: " + targetClasses.toString().replaceAll("/", ".")) +
+                            "\n\t\tat " + elementAsString,
+                    mixinClassName
+            );
```

Comment:
`blameIfPresent(mixinClassName, () -> blah)`
Making it a supplier, allows you to lazily resolve the string, which means that in the cases where there is no matching base package, you don't compute the big string for nothing.